File: examples/ch11/ts/conceptAdapter.ts
// Adapter — canonical conceptual example. Run: npx tsx conceptAdapter.ts
//
// A Target interface the client expects, an Adaptee with an incompatible API
// (typically code you don't own), and an Adapter that implements Target by
// translating to the Adaptee.

// --- Canonical: an Adapter wrapping an incompatible Adaptee -----------------

export interface Target {
  request(): string;
}

export class Adaptee {
  // Existing/foreign class with an incompatible interface.
  specificRequest(): string {
    return "data-from-adaptee";
  }
}

export class Adapter implements Target {
  constructor(private readonly adaptee: Adaptee) {}
  request(): string {
    const raw = this.adaptee.specificRequest();
    return `translated(${raw})`; // the real work: translate to Target's shape
  }
}

export const client = (target: Target): string => target.request();

// --- Structural: a class already shaped like Target needs no adapter --------
// TS types are structural, so any object with a matching `request()` IS a Target.

export class AlreadyShaped {
  request(): string {
    return "native";
  }
}

// Demo
console.log(client(new Adapter(new Adaptee()))); // translated(data-from-adaptee)
console.log(client(new AlreadyShaped())); // native — satisfies Target by shape, no wrapper

File: examples/ch11/py/concept_adapter.py
"""Adapter — canonical conceptual example (runnable).

A Target interface the client expects, an Adaptee with an incompatible API
(typically code you don't own), and an Adapter that implements Target by
translating to the Adaptee.
"""

from __future__ import annotations

from abc import ABC, abstractmethod


class Target(ABC):
    @abstractmethod
    def request(self) -> str: ...


class Adaptee:
    """Existing/foreign class with an incompatible interface."""

    def specific_request(self) -> str:
        return "data-from-adaptee"


class Adapter(Target):
    def __init__(self, adaptee: Adaptee) -> None:
        self._adaptee = adaptee

    def request(self) -> str:
        raw = self._adaptee.specific_request()
        return f"translated({raw})"   # the real work: translate to Target's shape


def client(target: Target) -> str:
    return target.request()           # the client only knows Target


if __name__ == "__main__":
    print(client(Adapter(Adaptee())))  # translated(data-from-adaptee)
